File: src/main/java/com/back/domain/game/gameLike/service/GameLikeService.java
package com.back.domain.game.gameLike.service;

import com.back.domain.game.game.entity.Game;
import com.back.domain.game.game.repository.GameRepository;
import com.back.domain.game.gameLike.entity.GameLike;
import com.back.domain.game.gameLike.repository.GameLikeRepository;
import com.back.domain.game.recommendation.event.ProfileVectorUpdateEvent;
import com.back.domain.member.member.entity.Member;
import com.back.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class GameLikeService {
    private final GameLikeRepository gameLikeRepository;
    private final GameRepository gameRepository;
    private final ApplicationEventPublisher eventPublisher;

    @Transactional
    public boolean toggleLike(Member member, long igdbId) {
        Game game = gameRepository.findByIgdbId(igdbId)
                .orElseThrow(() -> new ServiceException("404-1", "게임을 찾을 수 없습니다. igdbId" + igdbId));

        Optional<GameLike> existingLike = gameLikeRepository.findByMemberAndGame(member, game);

        boolean liked;
        if (existingLike.isPresent()) {
            // 이미 있으면 취소
            gameLikeRepository.delete(existingLike.get());
            gameRepository.decrementLikeCount(game.getId());
            liked = false;
        } else {
            // 없으면 추가
            GameLike gameLike = GameLike.createGameLike(member, game);
            gameLikeRepository.save(gameLike);
            gameRepository.incrementLikeCount(game.getId());
            liked = true;
        }
        eventPublisher.publishEvent(new ProfileVectorUpdateEvent(member.getId(), "toggleLike"));
        return liked;
    }

    @Transactional(readOnly = true)
    public boolean isLiked(Member member, long igdbId) {
        if (member == null) return false;
        return gameLikeRepository.existsByMemberIdAndGameIgdbId(member.getId(), igdbId);
    }

    @Transactional(readOnly = true)
    public long getLikeCount(long igdbId) {
        return gameRepository.findByIgdbId(igdbId)
                .map(Game::getLikeCount)
                .orElse(0L);
    }

//    @Transactional(readOnly = true)
//    public List<Game> getLikedGames(Member member) {
//        return gameLikeRepository.findLikedGamesByMember(member);
//    }
}

File: src/main/kotlin/com/back/domain/game/gameLike/service/GameLikeService.kt
package com.back.domain.game.gameLike.service

import com.back.domain.game.game.repository.GameRepository
import com.back.domain.game.gameLike.entity.GameLike
import com.back.domain.game.gameLike.repository.GameLikeRepository
import com.back.domain.game.recommendation.event.ProfileVectorUpdateEvent
import com.back.domain.member.member.entity.Member
import com.back.global.exception.ServiceException
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class GameLikeService(
    private val gameLikeRepository: GameLikeRepository,
    private val gameRepository: GameRepository,
    private val eventPublisher: ApplicationEventPublisher,
) {
    @Transactional
    fun toggleLike(
        member: Member,
        igdbId: Long,
    ): Boolean {
        val game =
            gameRepository
                .findByIgdbId(igdbId)
                .orElseThrow { ServiceException("404-1", "게임을 찾을 수 없습니다. igdbId$igdbId") }

        val existingLike = gameLikeRepository.findByMemberAndGame(member, game)

        val liked: Boolean
        if (existingLike.isPresent) {
            // 이미 있으면 취소
            gameLikeRepository.delete(existingLike.get())
            gameRepository.decrementLikeCount(game.id!!)
            liked = false
        } else {
            // 없으면 추가
            gameLikeRepository.save(GameLike.createGameLike(member, game))
            gameRepository.incrementLikeCount(game.id!!)
            liked = true
        }
        eventPublisher.publishEvent(ProfileVectorUpdateEvent(member.id, "toggleLike"))
        return liked
    }

    @Transactional(readOnly = true)
    fun isLiked(
        member: Member?,
        igdbId: Long,
    ): Boolean {
        if (member == null) return false
        return gameLikeRepository.existsByMemberIdAndGameIgdbId(member.id, igdbId)
    }

    @Transactional(readOnly = true)
    fun getLikeCount(igdbId: Long): Long =
        gameRepository
            .findByIgdbId(igdbId)
            .map { it.likeCount }
            .orElse(0L)
}
